### ISSUE TITLE
Loki Output timestamp not parsed properly

### ISSUE THREAD

**Issue by Z18Aplha** [fluent/fluent-bit#3350]
## Bug Report
First of all: I use the docker container of fluent bit v1.7.3 (and also tested with 1.7.2).

**Description of the bug**
Sending any type of input (syslog, tail, forward) to loki ends in an error from loki refering to the timestamp.

**To Reproduce**
- create ```fluent-bit.conf``` with the content
```
[SERVICE]
    Flush        5
    Daemon       Off
    Log_Level    trace
    Parsers_File parsers.conf
    Plugins_File plugins.conf
    HTTP_Server  Off

[INPUT]
    Name forward
    Listen 0.0.0.0
    Port 24224

[OUTPUT]
    Name loki
    Match *
    Host loki
    Port 3100
    Labels job=docker-forward, agent=fluentbit

[OUTPUT]
    Name stdout
    Match *
```
- run loki and fluent-bit docker containers with docker-compose:
```
version: "3"

networks:
  loki:

services:
  loki:
    image: grafana/loki:2.2.1
    container_name: loki
    ports:
      - "3100:3100"
    command: -config.file=/etc/loki/local-config.yaml
    networks:
      - loki
   
  fluentbit:
    image: fluent/fluent-bit:1.7.3
    container_name: fluentbit
    ports:
      - "24224:24224"
    volumes:
      - ./fluent-bit.conf:/fluent-bit/etc/fluent-bit.conf
    command: /fluent-bit/bin/fluent-bit -c /fluent-bit/etc/fluent-bit.conf
    networks:
      - loki
```
- trigger a log message via the fluentd log driver implemented in docker with
```
docker run --rm --log-driver=fluentd -t ubuntu echo "Testing a log message"
```
- Now you should see something like
```
[0] 98bb7a08ba8b: [1617978084.000000000, {"container_name"=>"/zen_franklin", "source"=>"stdout", "log"=>"Testing a log message", "container_id"=>"98bb7a08ba8b403167488f8b6b9f3ea1dec3a0180d4a7b10753be88b0329daee"}]
```
and the error from loki
```
[2021/04/09 14:21:29] [error] [output:loki:loki.0] loki:3100, HTTP status=400 loghttp.PushRequest.Streams: []*loghttp.Stream: loghttp.Stream.Entries: error reading entry timestamp: strconv.ParseInt: parsing "18446744072487954432": value out of range, error found in #10 byte of ...|487954432","{\"conta|..., bigger context ...|nt":"fluentbit"},"values":[["18446744072487954432","{\"container_name\":\"/zen_franklin\",\"source\"|...
```
Loki complains about the parsing of the timestamp. In the log above the timestamp is correct and in the format of fluentbit (```SECONDS.NANOSECONDS```). Apparently loki expects a nanoseconds-timestamp as an integer so the parsing results in a wrong number ```18446744072487954432``` and decoding this timestamp to ```Monday, 22. Juli 2554 01:34:32.487 GMT+02:00 DST``` which is obviously *out of range*.

**Expected behavior**
Sending to loki causes no error and the timestamp is correct.

**Environment**
**Version used:**
- loki v2.2.1
- fluentbit v1.7.3

**Configuration:**
- see above

**Environment name and version:**
- Docker v18.09.1

**Filters and plugins:**
- just used the docker images as it comes

**Additional context**
I want to collect the logs of all my docker containers and monitor them with grafana. So i use Loki as the log storage. 
I also tried to use promtail for log collection and it worked but the CPU load was too high and i want to use fluentbit because its lightweight and written in C.
My host is a raspberrypi 4 (armv7) with 2GB of RAM as my docker host. I could test it on my Synology NAS (x86_64) but my production system is the raspi.

**Comment by nokute78** [comment]
Hmm, I can't reproduce it. My host is Ubuntu 20.04.2 (x86_64) and docker version is 19.03.8...
Loki plugin has own timestamp converting function. It may not work on raspi4.

https://github.com/fluent/fluent-bit/blob/1e8c00240903ebc71a9092313fd8b620e10632bc/plugins/out_loki/loki.c#L702-L717


## my testing result

```
fluentbit    | [2021/04/10 00:08:32] [trace] [input:forward:forward.0 at plugins/in_forward/fw.c:97] new TCP connection arrived FD=27
fluentbit    | [2021/04/10 00:08:32] [trace] [input:forward:forward.0 at plugins/in_forward/fw_conn.c:73] read()=176 pre_len=0 now_len=176
fluentbit    | [2021/04/10 00:08:32] [trace] [input:forward:forward.0 at plugins/in_forward/fw_conn.c:84] fd=27 closed connection
fluentbit    | [2021/04/10 00:08:36] [debug] [task] created task=0x7f0a8ac30980 id=0 OK
", "container_id"=>"689393755e15b6b748224dbc812084463a88626810a82ec5dd515c67712c093f"}]williams", "source"=>"stdout", "log"=>"testing a log message
fluentbit    | [2021/04/10 00:08:36] [debug] [http_client] not using http_proxy for header
fluentbit    | [2021/04/10 00:08:36] [debug] [http_client] header=POST /loki/api/v1/push HTTP/1.1
fluentbit    | Host: loki:3100
fluentbit    | Content-Length: 296
fluentbit    | User-Agent: Fluent-Bit
fluentbit    | Content-Type: application/json
fluentbit    | 
fluentbit    | 
fluentbit    | [2021/04/10 00:08:36] [debug] [out coro] cb_destroy coro_id=0
fluentbit    | [2021/04/10 00:08:36] [ info] [output:loki:loki.0] loki:3100, HTTP status=204
fluentbit    | [2021/04/10 00:08:36] [debug] [upstream] KA connection #27 to loki:3100 is now available
fluentbit    | [2021/04/10 00:08:36] [debug] [out coro] cb_destroy coro_id=0
fluentbit    | [2021/04/10 00:08:36] [debug] [task] destroy task=0x7f0a8ac30980 (task_id=0)
^CGracefully stopping... (press Ctrl+C again to force)
```

**Comment by nokute78** [comment]
@Z18Aplha 
* Did raspi4 send "Testing a log message" ?
* Did fluent-bit and loki run on raspi4 ?
* Can you compile and test this C code?

```c
#include <time.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

int main(void) {
  struct timespec t;
  int len;
  char buf[64];
  uint64_t nanosecs;

  t.tv_sec = 1617978084;
  t.tv_nsec = 0;

  /* convert to nanoseconds */ 
  nanosecs = ((t.tv_sec * 1000000000L) + t.tv_nsec); 
  
  /* format as a string */ 
  len = snprintf(buf, sizeof(buf) - 1, "%" PRIu64, nanosecs); 

  printf("len=%d\nbuf=%s\n", len, buf);

  return 0;
}
```

My result is 
```
taka@locals:~$ gcc a.c 
taka@locals:~$ ./a.out 
len=19
buf=1617978084000000000
```

**Comment by Z18Aplha** [comment]
@nokute78 
Thank you very much for your fast answer and effort!

The format of your testing result looks a little bit different to mine but i think thats no problem.

> Did raspi4 send "Testing a log message" ?

Yes, i use the same raspi4 to send the log message.
> Did fluent-bit and loki run on raspi4 ?

Yes, both of them are on my raspi4 in a seperate docker image. But i think you want to know whether it runs on raspi without docker? If so: I dont know. Is it worth to test? In my case - i dont like local instances and really want to avoid using them if i can.
> Can you compile and test this C code?

Yes i succesfully compiled it with 
```
len=20
buf=18446744072487954432
```

Can you confirm the 'testing a log message' appears in loki? At some point in my testing and trying i had a ```HTTP status 204``` too but i cant find the log message in loki. I use grafana to visualize my logs.

**Comment by Z18Aplha** [comment]
@nokute78 
Interesting news: I did everything now on my Synology DS918+ (Linux DiskStation 4.4.59+ x86_64) and it works as expected. I get an HTTP 204 and the log message appears in grafana. 
So it has something to do with the raspberry i would say.

**Comment by Z18Aplha** [comment]
oh damn.. i recognize the timestamp in 
```
len=20
buf=18446744072487954432
```
is wrong. Its the same "out of range" timestamp loki complains about. So the pack_timestamp functions seems to be the problem.

**Comment by Z18Aplha** [comment]
Well. I'm not sure if this solution is universal for all platforms but with casting the t.tv_sec to an ```uint64_t``` it works:
```
/* convert to nanoseconds */
nanosecs = (((uint64_t)t.tv_sec * 1000000000L) + t.tv_nsec);
```
So my output matches yours:
```
len=19
buf=1617978084000000000
```
But i have still one question.. why it works without this cast on your machine but not on raspberrypi? Has it to do with 32bit/64bit?

**Comment by nokute78** [comment]
@Z18Aplha Thank you for information.

I can reproduce. It is an issue on 32bit platform.

```
taka@locals:~$ gcc -m32 a.c 
taka@locals:~$ ./a.out 
len=20
buf=18446744072487954432
```

> It seems raspberrypi uses 4Byte (instead of 8) for the time_t of t.tv_sec.

Yes. time_t is dependent of 32bit/64bit. It causes Year 2038 problem.

**Comment by nokute78** [comment]
I sent a patch https://github.com/fluent/fluent-bit/pull/3351

@Z18Aplha You can build it like following command.

```
git clone https://github.com/nokute78/fluent-bit.git
cd fluent-bit/build
git checkout fix_3350
cmake .. && make
bin/fluent-bit -i dummy -o loki
```

**Comment by nokute78** [comment]
@Z18Aplha The patch has been merged to master, so this issue is closed.

**Comment by Z18Aplha** [comment]
Thank you very much! You know when the docker image provides the patched version?

**Comment by nokute78** [comment]
I don't know when.
The patched version will be v1.7.4 or later.

**Comment by Z18Aplha** [comment]
Ok thank you!
